File: will-tan/week_01/MTA/MTA.js
// Objects declaration: each train line is an object with key 'stopName' that stores an array of names of stops on the train line
var lineL = {
	lineName: 'L line',
	stopName: [ '8th On L', '6th', 'Union Square', '3rd', '1st' ]
};

var lineSixth = {
	lineName: '6 line',
	stopName: [ 'Grand Central', '33rd', '28th On 6', '23rd On 6', 'Union Square', 'Astor']
};

var lineN = {
	lineName: 'N line',
	stopName: [ 'Time Square', '34th', '28th On N', '23rd On N', 'Union Square', '8th On N']
};

// Print out train trip, stop by stop
var planTrip = function ( startLine, startStop, endLine, endStop ) {
	var numStop = 0;
	var startLineStartStopIndex = startLine.stopName.indexOf( startStop );
	var startLineEndStopIndex = startLine.stopName.indexOf( endStop );
	var startLineUSquareIndex = startLine.stopName.indexOf( 'Union Square' );
	var endLineUSquareIndex = endLine.stopName.indexOf( 'Union Square' );
	var endLineEndStopIndex = endLine.stopName.indexOf( endStop );
	var announcement;

	if ( startLine === endLine ) { // check if user is going to the same train line
		if ( startStop === endStop ) { // check if user is already at his destination	
			console.log( "You are at your destination." );
		} else {
			announcement = printStops( startLine, startLineStartStopIndex, startLineEndStopIndex );
			console.log( "You must travel through the following stops on the " + startLine.lineName + ": " + announcement );
			printNumOfStops( startLineStartStopIndex, startLineEndStopIndex );
		}
	} else { // if going to different train line, get to Union Square first, then change
		if ( endStop === 'Union Square' ) { // check if user is going to Union Square
			announcement = printStops( startLine, startLineStartStopIndex, startLineUSquareIndex );
			console.log( "You must travel through the following stops on the " + startLine.lineName + ": " + announcement );
			printNumOfStops( startLineStartStopIndex, startLineUSquareIndex );
		} else if ( startStop === 'Union Square' ) {
			announcement = printStops( endLine, endLineUSquareIndex, endLineEndStopIndex );
			console.log("You must travel through the following stops on the " + endLine.lineName + ": " + announcement );
			printNumOfStops( startLineStartStopIndex, startLineUSquareIndex, endLineUSquareIndex, endLineEndStopIndex );
		} else {
			announcement = printStops( startLine, startLineStartStopIndex, startLineUSquareIndex );
			console.log( "You must travel through the following stops on the " + startLine.lineName + ": " + announcement );
			console.log( "Change at Union Square." );
			announcement = printStops( endLine, endLineUSquareIndex, endLineEndStopIndex );
			console.log( "Your journey continues through the following stops: " + announcement);
			printNumOfStops( startLineStartStopIndex, startLineUSquareIndex, endLineUSquareIndex, endLineEndStopIndex );
		}
	}
};

// Print out the stops, given the train line the user is on and the indices of the start and end stops of that train line
var printStops = function ( startLine, startIndex, endIndex ) {
	var str = '';

	if ( startIndex < endIndex ) { // if end stop is after start stop, travel forward and store each stop name in str
		for ( var i = startIndex + 1; i <= endIndex; i++ ) {
			if ( i === endIndex ) {
				str += startLine.stopName[ i ] + '.';
			}
			else {
				str += startLine.stopName[ i ] + ', ';
			}
		}
	} else { // travel backward
		for ( var i = startIndex - 1; i >= endIndex; i-- ) {
			if ( i === endIndex ) {	
				str += startLine.stopName[ i ] + '.';
			}
			else {
				str += startLine.stopName[ i ] + ', ';
			}
		}
	}
	return str;
};

var printNumOfStops = function ( startLineStartStopIndex, startLineEndStopIndex, endLineStartStopIndex, endLineEndStopIndex ) {
	var endLineStartStopIndex = endLineStartStopIndex || 0;
	var endLineEndStopIndex = endLineEndStopIndex || 0;
	var numStops = 0;
	numStops = Math.abs( startLineStartStopIndex - startLineEndStopIndex ); // calculate number of stops
	numStops += Math.abs( endLineStartStopIndex - endLineEndStopIndex );
	console.log( numStops + " stops in total." );
};

/************ test cases for same line ************/

console.log( "test case 1: same line forward" ); 
planTrip( lineSixth, 'Grand Central', lineSixth, 'Astor' );
console.log( "" );

console.log( "test case 2: same line backward" );
planTrip( lineL, '1st', lineL, '8th On L' );
console.log( "" );

console.log( "test case 3: same line same stop" );
planTrip( lineL, '1st', lineL, '1st' );
console.log( "" );

/************ test cases for different line ************/

console.log( "test case 4: different line but stop at 'Union Square'" );
planTrip( lineL, '1st', lineSixth, 'Union Square' );
console.log( "" );

console.log( "test case 5: forward to Union Square, Union Square forward to destination" );
planTrip( lineN, 'Time Square', lineL, '1st' );
console.log( "" );

console.log( "test case 6: backward to Union Square, Union Square forward to destination" );
planTrip( lineN, '8th On N', lineSixth, 'Astor' );
console.log( "" );

console.log( "test case 7: forward to Union Square, Union Square backward to destination" );
planTrip( lineSixth, 'Grand Central', lineN, 'Time Square' );
console.log( "" );

console.log( "test case 8: backward to Union Square, Union Square backward to destination" );
planTrip( lineSixth, 'Astor', lineL, '8th On L' );
console.log( "" );

console.log("test case 9: start at Union Square, Union Square forward to destination" );
planTrip( lineL, 'Union Square', lineN, '8th On N' );
console.log( "" );

console.log("test case 10: start at Union Square, Union Square backward to destination" );
planTrip( lineSixth, 'Union Square', lineL, '8th On L' );
console.log( "" );

console.log("test case 11: same stop, different line" );
planTrip( lineN, '8th on N', lineL, '8th On L' );
console.log( "" );


		// My Own Train of Thoughts
		// I want to find the index of startStop and endStop
		// console.log( startLine.stopName.indexOf( startStop ) );

		// Then, print out every stop in between startStop and endStop
		// console.log( startLine.stopName.indexOf( endStop ) );

		// If endStop index is > startStop index, travel backwards; else, travel forward
		// if ( startLine.stopName.indexOf( startStop ) < startLine.stopName.indexOf( endStop ) ) {
		// 	// print the stopNames forward,
		// 	printStops( startLine, startLine.stopName.indexOf( startStop ), startLine.stopName.indexOf( endStop ) );
		// } else {
		// 	// print the stopNames backward
		// 	//printStops( startLine, indexOf( endStop ), indexOf( startStop ) );
		// }

File: kylie-obrien-pratt/week_04/Day_02/MTA.rb
# // MTA

# // // don't need to use constructors or factories for MTA. stick to arrays of strings
# // Find indexes of the start and finish stations
# // Aim to set up arrays and objects at minimum
# //  The program should work out which stops you need to go to like Trip Planner
# // Program should state which stations you need to go through and the total number of stations
# // You don't know what direction you're going to be going in - you could go forward or backwards or both
# // Union Square is where they change from one line to another line

# // You don't need to prompt the user for input, hard code planTrip = function()
# // Get program to work for one line before you tackle multiple lines

# // Tell the user the number of stops AND the stops IN ORDER that they will pass through or change at.
# // The key to the lab is finding the index positions of each stop. (hint: indexOf())

# // console.log() shows output similar to this:
# // "Your must travel through the following stops on the N line: 34th, 28th, 23rd, Union Square."
# // "Change at Union Square."
# // "Your journey continues through the following stops: 23rd, 28th, 33rd."
# // "7 stops in total."

	
	// can use startingLine.slice; startingLine.reverse (permanently reverses variable, therefore create new variable reverseLine = lines[arrivalLine].reverse
	// array.length console.log number of stops. "n stops in total"
	// use separate section to logic for prompting the user, makes it easier to debug and reuse, 
	// could still use it for applications where you don't prompt the user
//

method.abs #convert to absolute value

#newline.chomp


lineN = ["Time Square", "34th", "28th", "23rd", "Union Square", "8th"]
lineL = ["8th", "6th", "Union Square", "3rd", "1st"]
line6 = ["Grand Central", "33rd", "28th", "23rd", "Union Square", "Astor Place"]


def planTrip = (startLine, startStation, endLine, endStation) {
	//Check if the end station and line are the same as the starting station and line
	if(startLine == endLine && startStation == endStation) {
		console.log("The starting point is the same as the end point");
		return;
	}
	//Check to see if they are on the same line
	if(startLine == endLine) {
		var lineRef;
		if(startLine === "N") {
			lineRef = lineN;
		}
		else if(startLine === "L") {
			lineRef = lineL;
		}
		else if(startLine === "6") {
			lineRef = line6;
		}

		if(startStationIndex > endStationIndex) {
			console.log("Your journey will be as follows:");
			var indexDifference = startStation - endStationIndex;
			for(var i = startStation; i >= indexDifference; i--) {
				console.log(lineRef[i]);
			}
			
		}

		else {
			console.log("Your journey will be as follows:");
			var indexDifference = endStationIndex - startStation;
			for(var i = startStation; i <= indexDifference; i++) {
				console.log(lineRef[i]);
			}
		}
	}

	var endLineRef;
	if(startLine === "N") {
		startLineRef = lineN;
	}
	else if(startLine === "L") {
		startLineRef = lineL;
	}
	else if(startLine === "6") {
		startLineRef = line6;
	}
	else {
		console.log("Invalid line, valid lines are: N, L, 6");
		return;
	}

	var startStationIndex = startLineRef.indexOf(startStation);
	if(startStationIndex == NaN) {
		console.log("Station doesn't exist on starting line");
		return;
	}


	var endLineRef;
	if(endLine === "N") {
		endLineRef = lineN;
	}
	else if(endLine === "L") {
		endLineRef = lineL;
	}
	else if(endLine === "6") {
		endLineRef = line6;
	}
	else {
		console.log("Invalid line, valid lines are: N, L, 6");
		return;
	}
	
	var endStationIndex = endLineRef.indexOf(endStation);
	if(endStationIndex == NaN) {
		console.log("Station doesn't exist on the End Line");
		return;
		}
}
	//Find the path now


	//









# 	//

# 	lineN = [Times Square, 34th, 28th, 23rd, Union Square, 8th]
# lineL = [8th, 6th, Union Square, 3rd, 1st]
# line6 = [Grand Central, 33rd, 28th, 23rd, Union Square, Astor Place]

# def planTrip (startLine, startStation, endLine, endStation)
# 	# Check if the end station and line are the same as the starting station and line
# 	if(startLine == endLine && startStation == endStation) {
# 		print "The starting point is the same as the end point"
	

# # Check to see if they are on the same line

# 	if startLine == endLine 
		
# 		if startLine == "N"
# 			lineRef = lineN;
		
# 		elsif startLine == "L" 
# 			lineRef = lineL
		
# 		elsif startLine == "6"
# 			lineRef = line6;
# 	end

# 	if startStationIndex > endStationIndex 
# 		print "Your journey will be as follows: "
# 		# prints list of stations on journey
# 		def indexDifference = startStation - endStationIndex;
# 			i = startStation
# 			until == indexDifference
#     		puts "lineRef: #{ i }"
#     		i -= 1

# 	end

# 	else 
# 			print "Your journey will be as follows: "
# 			def indexDifference = endStationIndex - startStation
# 			i = startStation
# 			until == indexDifference
#     		puts "lineRef: #{ i }"
#     		i += 1
		
#     	end

# end

